File: frontend/src/lib/CFGGenerator.ts
// ============================================================
// CFGGenerator.ts — Control Flow Graph from LLVM IR
// ============================================================
import { CFGBlock, CFGEdge, CFGGraph, ParsedInstruction, Opcode } from "@/types"

interface RawBlock {
  id: string
  label: string
  lines: string[]
  instructions: ParsedInstruction[]
  successorLabels: string[]
}

export class CFGGenerator {
  private ir: string

  constructor(ir: string) {
    this.ir = ir
  }

  generate(): CFGGraph {
    const rawBlocks = this.parseBlocks()
    if (rawBlocks.length === 0) {
      // Single basic block (no labels) — trivial CFG
      return this.buildTrivialCFG()
    }
    return this.buildCFG(rawBlocks)
  }

  private parseBlocks(): RawBlock[] {
    const lines = this.ir.split("\n")
    const blocks: RawBlock[] = []
    let currentBlock: RawBlock | null = null
    let instId = 0

    for (const rawLine of lines) {
      const line = rawLine.trim()
      if (!line || line.startsWith(";")) continue
      if (line.startsWith("define") || line === "{" || line === "}") continue

      // Block label: ends with ":"
      const labelMatch = line.match(/^(\w+):/)
      if (labelMatch) {
        if (currentBlock) blocks.push(currentBlock)
        currentBlock = {
          id: labelMatch[1],
          label: labelMatch[1],
          lines: [],
          instructions: [],
          successorLabels: [],
        }
        continue
      }

      if (!currentBlock) {
        // Before any label — treat as "entry" block
        currentBlock = {
          id: "entry",
          label: "entry",
          lines: [],
          instructions: [],
          successorLabels: [],
        }
      }

      currentBlock.lines.push(line)

      // Parse instructions for this block
      const parsed = this.parseLine(line, instId, currentBlock.id)
      if (parsed) {
        currentBlock.instructions.push(parsed)
        instId++
      }

      // Detect branch terminators
      if (line.startsWith("br ")) {
        const targets = this.extractBranchTargets(line)
        currentBlock.successorLabels.push(...targets)
      } else if (line.startsWith("ret ") || line === "ret void") {
        // no successors
      }
    }

    if (currentBlock) blocks.push(currentBlock)
    return blocks
  }

  private parseLine(line: string, id: number, blockId: string): ParsedInstruction | null {
    const cleaned = line.replace(/;.*$/, "").trim()
    if (!cleaned) return null

    let result = ""
    let rest = cleaned
    const eqIdx = cleaned.indexOf("=")
    if (eqIdx !== -1) {
      result = cleaned.substring(0, eqIdx).trim()
      rest = cleaned.substring(eqIdx + 1).trim()
    }

    const opcode = rest.split(/\s+/)[0] as Opcode
    const validOpcodes: Opcode[] = ["add", "sub", "mul", "load", "store", "phi", "ret", "br", "icmp", "alloca"]
    if (!validOpcodes.includes(opcode)) return null

    const operands: string[] = []
    const tokens = rest.substring(opcode.length).split(/[\s,]+/)
    for (const t of tokens) {
      if (t.startsWith("%")) operands.push(t)
    }

    return { id, result, opcode, operands, text: line.trim(), blockId }
  }

  private extractBranchTargets(line: string): string[] {
    // br i1 %cond, label %true, label %false
    // br label %target
    const targets: string[] = []
    const labelMatches = line.matchAll(/label %(\w+)/g)
    for (const m of labelMatches) {
      targets.push(m[1])
    }
    return targets
  }

  private buildTrivialCFG(): CFGGraph {
    // Parse all instructions as a single entry block
    const lines = this.ir.split("\n")
    const instructions: ParsedInstruction[] = []
    let id = 0
    for (const raw of lines) {
      const line = raw.replace(/;.*$/, "").trim()
      if (!line) continue
      if (line.startsWith("define") || line.startsWith("}") || line.includes(":") && !line.startsWith("%")) continue
      const parsed = this.parseLine(line, id, "entry")
      if (parsed) { instructions.push(parsed); id++ }
    }

    const entryBlock: CFGBlock = {
      id: "entry",
      label: "entry (single block)",
      instructions,
      successors: [],
      predecessors: [],
      isEntry: true,
      isExit: true,
    }

    return {
      blocks: [entryBlock],
      edges: [],
      entry: "entry",
      exits: ["entry"],
    }
  }

  private buildCFG(rawBlocks: RawBlock[]): CFGGraph {
    const blockMap = new Map(rawBlocks.map(b => [b.id, b]))
    const edges: CFGEdge[] = []

    // Build edges
    for (const block of rawBlocks) {
      for (const succ of block.successorLabels) {
        if (blockMap.has(succ)) {
          edges.push({ source: block.id, target: succ, isBackEdge: false })
        }
      }
    }

    // Detect back edges (simple: if target appears before source in block order)
    const blockOrder = rawBlocks.map(b => b.id)
    for (const edge of edges) {
      const srcIdx = blockOrder.indexOf(edge.source)
      const tgtIdx = blockOrder.indexOf(edge.target)
      if (tgtIdx <= srcIdx) {
        edge.isBackEdge = true
        edge.label = "loop back"
      }
    }

    // Build predecessor map
    const predMap = new Map<string, string[]>()
    for (const b of rawBlocks) predMap.set(b.id, [])
    for (const edge of edges) {
      predMap.get(edge.target)?.push(edge.source)
    }

    // Detect loop headers (targets of back edges)
    const loopHeaders = new Set(edges.filter(e => e.isBackEdge).map(e => e.target))

    // Build CFGBlocks
    const cfgBlocks: CFGBlock[] = rawBlocks.map((raw, idx) => ({
      id: raw.id,
      label: raw.label,
      instructions: raw.instructions,
      successors: raw.successorLabels.filter(s => blockMap.has(s)),
      predecessors: predMap.get(raw.id) || [],
      isEntry: idx === 0,
      isExit: raw.successorLabels.length === 0,
      isLoopHeader: loopHeaders.has(raw.id),
    }))

    const exits = cfgBlocks.filter(b => b.isExit).map(b => b.id)

    return {
      blocks: cfgBlocks,
      edges,
      entry: rawBlocks[0]?.id || "entry",
      exits,
    }
  }
}

File: backend/src/CFGGenerator.cpp
#include "CFGGenerator.h"
#include <sstream>
#include <algorithm>

CFGGenerator::CFGGenerator(const std::string& ir) : ir(ir) {}

void CFGGenerator::generate() {
    parseBlocks();
    buildEdges();
}

void CFGGenerator::parseBlocks() {
    std::istringstream iss(ir);
    std::string line;
    int instId = 0;
    
    CFGBlock* currentBlock = nullptr;
    
    while (std::getline(iss, line)) {
        // Trim
        line.erase(0, line.find_first_not_of(" \t\r\n"));
        line.erase(line.find_last_not_of(" \t\r\n") + 1);
        if (line.empty() || line[0] == ';') continue;
        if (line.find("define") != std::string::npos || line == "{" || line == "}") continue;
        
        // Check for block label
        size_t colonPos = line.find(":");
        if (colonPos != std::string::npos && line.find("=") == std::string::npos && line[0] != '%') {
            std::string label = line.substr(0, colonPos);
            CFGBlock newBlock;
            newBlock.id = label;
            newBlock.label = label;
            blocks.push_back(newBlock);
            currentBlock = &blocks.back();
            continue;
        }
        
        if (!currentBlock) {
            CFGBlock newBlock;
            newBlock.id = "entry";
            newBlock.label = "entry";
            blocks.push_back(newBlock);
            currentBlock = &blocks.back();
        }
        
        // Parse instruction
        Instruction inst = Instruction::parse(line, instId);
        if (inst.opcode != Opcode::NOP || !inst.result.empty()) {
            inst.index = instId++;
            currentBlock->instructions.push_back(inst);
            
            // Detect branches
            if (line.find("br ") == 0) {
                size_t pos = 0;
                while ((pos = line.find("label %", pos)) != std::string::npos) {
                    pos += 7; // length of "label %"
                    size_t end = line.find_first_of(" ,", pos);
                    std::string target = line.substr(pos, end - pos);
                    currentBlock->successors.push_back(target);
                }
            }
        }
    }
    
    if (!blocks.empty()) {
        blocks.front().isEntry = true;
        entryBlockId = blocks.front().id;
    }
}

void CFGGenerator::buildEdges() {
    for (auto& block : blocks) {
        if (block.successors.empty()) {
            block.isExit = true;
            exitBlockIds.push_back(block.id);
        }
        for (const auto& succ : block.successors) {
            CFGEdge edge;
            edge.source = block.id;
            edge.target = succ;
            
            // Determine if back edge (target appears before source in block order)
            auto srcIt = std::find_if(blocks.begin(), blocks.end(), [&](const CFGBlock& b) { return b.id == block.id; });
            auto tgtIt = std::find_if(blocks.begin(), blocks.end(), [&](const CFGBlock& b) { return b.id == succ; });
            
            if (tgtIt <= srcIt) {
                edge.isBackEdge = true;
                edge.label = "loop back";
                if (tgtIt != blocks.end()) {
                    tgtIt->isLoopHeader = true;
                }
            }
            edges.push_back(edge);
            
            // Add predecessor
            if (tgtIt != blocks.end()) {
                tgtIt->predecessors.push_back(block.id);
            }
        }
    }
}

std::vector<Instruction> CFGGenerator::getAllInstructions() const {
    std::vector<Instruction> allInsts;
    for (const auto& b : blocks) {
        for (const auto& i : b.instructions) {
            allInsts.push_back(i);
        }
    }
    return allInsts;
}

std::string escapeJSON(const std::string& str) {
    std::string escaped;
    for (char c : str) {
        if (c == '"') escaped += "\\\"";
        else if (c == '\\') escaped += "\\\\";
        else escaped += c;
    }
    return escaped;
}

std::string CFGGenerator::toJSON() const {
    std::string json = "{\"blocks\":[";
    for (size_t i = 0; i < blocks.size(); ++i) {
        const auto& b = blocks[i];
        json += "{\"id\":\"" + b.id + "\",\"label\":\"" + b.label + "\",\"isEntry\":" + (b.isEntry ? "true" : "false") + 
                ",\"isExit\":" + (b.isExit ? "true" : "false") + ",\"isLoopHeader\":" + (b.isLoopHeader ? "true" : "false") + 
                ",\"instructions\":[";
        for (size_t j = 0; j < b.instructions.size(); ++j) {
            json += "{\"id\":" + std::to_string(b.instructions[j].index) + ",\"text\":\"" + escapeJSON(b.instructions[j].toString()) + "\"}";
            if (j < b.instructions.size() - 1) json += ",";
        }
        json += "]}";
        if (i < blocks.size() - 1) json += ",";
    }
    json += "],\"edges\":[";
    for (size_t i = 0; i < edges.size(); ++i) {
        const auto& e = edges[i];
        json += "{\"source\":\"" + e.source + "\",\"target\":\"" + e.target + "\",\"isBackEdge\":" + (e.isBackEdge ? "true" : "false") + "}";
        if (i < edges.size() - 1) json += ",";
    }
    json += "]}";
    return json;
}
